Alerting: Add manage permissions UI logic for Contact Points

tomratcliffe opened the pull request:
**What is this feature?**

Adds capability to manage permissions to read/write/admin individual contact points on the Grafana alertmanager, when `accesscontrol` license feature is enabled

**Why do we need this feature?**

More granular control over access to contact points, so it's not all or nothing for access to contact points.

**Who is this feature for?**

Alerting users who have `accesscontrol` license feature enabled, and `alertingApiServer` feature toggle enabled

tomratcliffe committed:
Add showPolicies prop

tomratcliffe committed:
Add manage permissions component for easier reuse within alerting

tomratcliffe committed:
Add method for checking whether to show access control within alerting

tomratcliffe committed:
Remove accidental console.log from main

tomratcliffe committed:
Tweak styling for contact point width and add manage permissions drawer

tomratcliffe committed:
Improve typing for access control type response

tomratcliffe committed:
Add basic test for manage permissions on contact points list

tomratcliffe committed:
Only show manage permissions if grafana AM and alertingApiServer is enabled

tomratcliffe committed:
Update i18n

tomratcliffe committed:
Add test utils for turning features on and back off

tomratcliffe committed:
Add access control handlers

tomratcliffe committed:
Update tests with new util

tomratcliffe committed:
Pass AM in and add tests

JacobsonMT committed:
Receiver OSS resource permissions

There is a complication that is not fully addressed: Viewer defaults to read:*
and Editor defaults to read+write+delete:*

This is different to other resource permissions where non-admin are not granted
any global permissions and instead access is handled solely by resource-specific
permissions that are populated on create and removed on delete.

This allows them to easily remove permission to view or edit a single resource
from basic roles.

The reason this is tricky here is that we have multiple APIs that can
create/delete receivers: config api, provisioning api, and k8s receivers api.
Config api in particular is not well-equipped to determine when creates/deletes
are happening and thus ensuring that the proper resource-specific permissions
are created/deleted is finicky.

We would also have to create a migration to populate resource-specific
permissions for all current receivers. This migration would need to be reset so
it can run again if the flag is disabled.

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

tomratcliffe committed:
Add access control permissions

tomratcliffe committed:
Pass in contact point ID to receivers form

tomratcliffe committed:
Temporarily remove access control check for contact points

JacobsonMT committed:
Include access control metadata in k8s receiver List & Get

GET: Always included.
LIST: Included by adding a label selector with value `grafana.com/accessControl`

JacobsonMT committed:
Include in-use metadata in k8s receiver List & Get

GET: Always included.
LIST: Included by adding a label selector with value `grafana.com/inUse`

JacobsonMT committed:
Include new permissions for contact points navbar

JacobsonMT committed:
Fix receiver creator fixed role to not give global read

JacobsonMT committed:
Add receiver creator permission to receiver writer

JacobsonMT committed:
Add receiver creator permission to navbar

JacobsonMT committed:
Always allow listing receivers, don't return 403

JacobsonMT committed:
Remove receiver read precondition from receiver create

Otherwise, Creator role will not be able to create their first receiver

tomratcliffe committed:
Update routes permissions

tomratcliffe committed:
Add further support for RBAC in contact points

tomratcliffe committed:
Merge branch 'main' into alerting/contact-point-permissions

tomratcliffe commented:
/deploy-to-hg

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -514,6 +574,39 @@ func (rs *ReceiverService) checkOptimisticConcurrency(receiver *models.Receiver,
 	return nil
 }
 
+func (rs *ReceiverService) setDefaultPermissions(ctx context.Context, orgID int64, user identity.Requester, receiver models.Receiver) {
```
I wonder if it should be part of our access control (`autz`) service?

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

tomratcliffe committed:
Update routes permissions

tomratcliffe committed:
Update contact points header logic

tomratcliffe committed:
Back out test feature toggle refactor

Not working atm, not sure why

tomratcliffe committed:
Tidy up imports

tomratcliffe committed:
Update mock permissions

tomratcliffe committed:
Revert more test changes

tomratcliffe committed:
Update i18n

JacobsonMT committed:
Merge branch 'main' into alerting/contact-point-permissions

# Conflicts:
#	pkg/apis/alerting_notifications/v0alpha1/types_ext.go
#	pkg/registry/apis/alerting/notifications/receiver/conversions.go
#	pkg/registry/apis/alerting/notifications/receiver/legacy_storage.go
#	pkg/registry/apis/alerting/notifications/register.go
#	pkg/services/ngalert/accesscontrol/receivers.go
#	pkg/services/ngalert/models/permissions.go

JacobsonMT committed:
Sync inuse metadata pr

JacobsonMT committed:
Add back canAdmin permissions after main merge

JacobsonMT committed:
Move SetDefaultPermissions to ReceiverPermissionService

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

tomratcliffe committed:
Split out check for policies navtree item

tomratcliffe committed:
Tidy up utils and imports and fix rules in use

tomratcliffe committed:
Fix contact point tests and act warnings

JacobsonMT committed:
Add missing ReceiverPermissionAdmin after merge conflict

JacobsonMT committed:
Merge branch 'main' into alerting/contact-point-permissions

# Conflicts:
#	pkg/services/ngalert/notifier/legacy_storage/receivers.go

JacobsonMT committed:
Have SetDefaultPermissions use uid from string

Fixes circular dependency

tomratcliffe committed:
Move contact points permissions

tomratcliffe committed:
Only show contact points filter when permissions are correct

tomratcliffe committed:
Move to constants

tomratcliffe committed:
Fallback to empty array and remove labelSelectors (not needed)

tomratcliffe committed:
Allow `toAbility` to take multiple actions

tomratcliffe committed:
Show builtin alertmanager if contact points permission

tomratcliffe committed:
Add empty state and hide templates if missing permissions

tomratcliffe committed:
Translations

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPointHeader.tsx:
```diff
@@ -107,10 +188,20 @@ export const ContactPointHeader = ({ contactPoint, disabled = false, onDelete }:
             {referencedByPoliciesText}
           </TextLink>
         )}
+        {numberOfRules > 0 && (
```
Now the `unused` badge for Grafana AM contact points should be accurate 🙌 As we can now distinguish from an auto generated simpified routing policy vs actually being used directly by a rule

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.test.tsx:
```diff
@@ -428,5 +449,28 @@ describe('contact points', () => {
+    it('does not show manage permissions', async () => {
+      renderGrafanaContactPoints();
+
+      await clickMoreActionsButton('lotsa-emails');
```
I initially forgot to `await` this click and it was causing all kinds of annoyances in the tests 🙃 
Created https://github.com/grafana/grafana/issues/93405 to enable this

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -41,10 +46,16 @@ const ContactPointsTab = () => {
+  // as we get metadata about this from the API
+  const fetchPolicies = !shouldUseK8sApi(selectedAlertmanager!);
+  // User may have access to list contact points, but not permission to fetch the status endpoint
+  const fetchStatuses = contextSrv.hasPermission(AccessControlAction.AlertingNotificationsRead);
```
Without this, a user with basic permissions would end up calling for contact points statuses and get an error toast on screen

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -58,16 +69,32 @@ const ContactPointsTab = () => {
 
   const isGrafanaManagedAlertmanager = selectedAlertmanager === GRAFANA_RULES_SOURCE_NAME;
+
+  if (contactPoints.length === 0) {
```
I don't think it was possibly/very likely at all to see an empty state before, as there was a default contact point, and it was tricky to end up deleting everything

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -99,7 +126,8 @@ const ContactPointsTab = () => {
         </Stack>
       </Stack>
-      <ContactPointsList contactPoints={contactPoints} search={search} pageSize={DEFAULT_PAGE_SIZE} />
+      {error && <Alert title="Failed to fetch contact points">{stringifyErrorLike(error)}</Alert>}
```
Had to move the error so we still render action/filters list for the case where permissions don't let you see the list

soniaAguilarPeiron replied:
I believe we should still keep the error outside as it was, as in case of error, whats the point on seeing the action/filters ?

tomratcliffe replied:
I moved it here as I was trying to avoid the possibility of any other error stopping the rendering of the actions - at the time of writing, it was possible for the list endpoint to return a 403, just because you couldn't _see_ any contact points (but you had permission to create them). Are you okay if I leave it this way, to mitigate the risk of that kind of thing (unexpectedly) happening?

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -159,6 +187,8 @@ const ContactPointsPageContents = () => {
     alertmanager: selectedAlertmanager!,
   });
 
+  const showTemplatesTab = contextSrv.hasPermission(AccessControlAction.AlertingNotificationsRead);
```
User could now have permissions for contact points, but not templates tab

tomratcliffe commented on public/app/features/alerting/unified/components/permissions/ManagePermissions.tsx:
There's not anything alerting specific in here at the moment, so it _could_ be used more widely? Just a more opinionated way to render a button/drawer on the page for permissions. We can fairly easily use this to manage folder permissions within alerting without having to go to dashboards (but there are UX problems with a modal + drawer, and I didn't want to try and solve that here)

tomratcliffe commented on public/app/features/alerting/unified/components/receivers/EditReceiverView.tsx:
```diff
@@ -17,7 +17,6 @@ export const EditReceiverView = ({ contactPoint, alertmanagerName }: Props) => {
   const readOnly = !editSupported || !editAllowed;
 
   if (alertmanagerName === GRAFANA_RULES_SOURCE_NAME) {
-    console.log(contactPoint);
```
Oops

tomratcliffe commented on public/app/features/alerting/unified/components/rules/Filter/RulesFilter.v1.tsx:
```diff
@@ -150,7 +152,10 @@ const RulesFilter = ({ onClear = () => undefined }: RulesFilerProps) => {
   };
 
-  const canRenderContactPointSelector = config.featureToggles.alertingSimplifiedRouting ?? false;
+  const canRenderContactPointSelector =
```
Otherwise we would render the filter selector when user potentially couldn't see it

konrad147 replied:
This is an interesting edge case. If I understand correctly, if the user has no `AlertingReceiversRead` permission we cannot use simplify routing and we need to default to notification policies?

tomratcliffe replied:
Yes - but also I've been thinking about edge cases of them being able to see contact points, but not necessarily the contact point associated with an alert rule they can see. So perhaps in that case we should disable the simplified routing selector?

tomratcliffe commented on public/app/features/alerting/unified/hooks/useAbilities.ts:
```diff
@@ -218,7 +218,12 @@ export function useAllAlertmanagerAbilities(): Abilities<AlertmanagerAction> {
+      hasConfigurationAPI,
+      notificationsPermissions.create,
+      // TODO: Move this into the permissions config and generalise that code to allow for an array of permissions
+      isGrafanaFlavoredAlertmanager ? AccessControlAction.AlertingReceiversCreate : null
```
We have some cases where it isn't just one permission that allows you to see something. This felt like the cleanest way to allow checking multiple access control actions without changing a lot of existing objects to be arrays with a single element in

tomratcliffe commented on public/app/features/alerting/unified/mocks/server/handlers/k8s/receivers.k8s.ts:
```diff
@@ -20,7 +21,12 @@ const mappedReceivers =
         // This isn't exactly accurate, but its the cleanest way to use the same data for AM config and K8S responses
         uid: camelCase(contactPoint.name),
-        annotations: { [PROVENANCE_ANNOTATION]: provenance },
+        annotations: {
```
TODO as well (maybe in this PR, maybe not, thoughts?) - a cleaner way to manage entities here and allow for contact points that have different permutations of permissions, and inUse information - e.g ones that are in use by alert rules but not policies and vice versa

tomratcliffe commented on public/app/features/alerting/unified/utils/datasource.ts:
```diff
@@ -147,7 +148,11 @@ export function getAlertManagerDataSourcesByPermission(permission: 'instance' |
   };
 
-  const builtinAlertmanagerPermissions = Object.values(permissions).flatMap((permissions) => permissions.grafana);
+  const builtinAlertmanagerPermissions = [
+    ...Object.values(permissions).flatMap((permissions) => permissions.grafana),
+    ...PERMISSIONS_CONTACT_POINTS,
+  ];
```
Needed to show the Grafana dropdown option if you gave a user _only_ contact point creation permissions - as otherwise they wouldn't see the option in the dropdown and the page would be blocked from rendering by the alert saying "you don't have any AMs"

konrad147 replied:
I wonder if we shouldn't include it in the `permissions` object, but I'd need to dive a bit deeper into the  code

tomratcliffe replied:
Are you okay if I leave it as is for now?

tomratcliffe committed:
Tidy up mock data

tomratcliffe committed:
Fix tests and templates permission

tomratcliffe committed:
Merge branch 'main' into alerting/contact-point-permissions

yuri-tceretian commented on pkg/services/accesscontrol/ossaccesscontrol/receivers.go:
```diff
@@ -0,0 +1,114 @@
+var ReceiversAdminActions = append(ReceiversEditActions, []string{accesscontrol.ActionAlertingReceiversReadSecrets, accesscontrol.ActionAlertingReceiversPermissionsRead, accesscontrol.ActionAlertingReceiversPermissionsWrite}...)
+
+func registerReceiverRoles(cfg *setting.Cfg, service accesscontrol.Service) error {
+	if !cfg.RBAC.PermissionsWildcardSeed("receiver") { // TODO: Do we need wildcard seed support in alerting?
```
I think it makes sense to follow the same pattern as other resources such as folder and datasources do.

yuri-tceretian commented on pkg/services/accesscontrol/ossaccesscontrol/receivers.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	options := resourcepermissions.Options{
+		Resource:          "receivers",
```
this should be `alertingac.ScopeReceiversRoot`, right?

yuri-tceretian commented on pkg/services/accesscontrol/accesscontrol.go:
```diff
@@ -148,6 +149,12 @@ type ServiceAccountPermissionsService interface {
 	PermissionsService
 }
 
+type ReceiverPermissionsService interface {
```
I wonder if this interface should be defined on user side in ngalert?

yuri-tceretian commented on pkg/services/navtree/navtreeimpl/navtree.go:
```diff
@@ -405,11 +405,23 @@ func (s *ServiceImpl) buildAlertNavLinks(c *contextmodel.ReqContext) *navtree.Na
-	if hasAccess(ac.EvalAny(ac.EvalPermission(ac.ActionAlertingNotificationsRead), ac.EvalPermission(ac.ActionAlertingNotificationsExternalRead))) {
+	if hasAccess(ac.EvalAny(
+		ac.EvalPermission(ac.ActionAlertingNotificationsRead),
+		ac.EvalPermission(ac.ActionAlertingNotificationsExternalRead),
```
I think we should still support legacy actions here. Otherwise, enterprise users that provision custom roles will lose access unless they migrate.

yuri-tceretian commented on pkg/services/ngalert/accesscontrol/receivers.go:
```diff
@@ -199,6 +231,18 @@ func NewReceiverAccess[T models.Identified](a ac.AccessControl, includeProvision
+			},
+			resource:      "receiver",
+			action:        "admin", // Essentially read+write receiver resource permissions.
+			authorizeSome: permissionsReceiversPreConditionsEval,
```
should it require `receivers:read` in the same scope?

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

tomratcliffe committed:
Update message for unused contact points

JacobsonMT committed:
Add FakeReceiverPermissionsService and fix test wiring

JacobsonMT committed:
Implement resource permission handling in provisioning API and renames

Create: Sets to default permissions
Delete: Removes permissions
Update: If receiver name is modified and the new name doesn't exist, it copies
the permissions from the old receiver to the newly created one. If old receiver
is now empty, it removes the old permissions as well.

JacobsonMT committed:
Don't return 403 when user lists receivers and has access to none

JacobsonMT committed:
Fix receiver create not adding empty uid permissions

JacobsonMT committed:
Proxy RouteDeleteAlertingConfig through MultiOrgAlertmanager

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

soniaAguilarPeiron commented on public/app/features/alerting/routes.tsx:
```diff
@@ -109,6 +111,7 @@ export function getAlertingRoutes(cfg = config): RouteDescriptor[] {
       roles: evaluateAccess([
         AccessControlAction.AlertingNotificationsWrite,
         AccessControlAction.AlertingNotificationsExternalWrite,
+        ...PERMISSIONS_CONTACT_POINTS,
```
With this addition, I 'viewer' user can go to this route (tested locally)

tomratcliffe replied:
Yes good catch, thank you! I've split the permissions now and pushed a change

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/ContactPointHeader.tsx:
```diff
@@ -26,21 +34,54 @@ interface ContactPointHeaderProps {
+   * either taken from the annotations (k8s) or from the alertmanager config
+   */
+  const numberOfPolicies = Number(
+    contactPoint.metadata?.annotations?.[ANNOTATION_INUSE_ROUTES] ?? regularPolicyReferences.length
```
🎉

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/ContactPointHeader.tsx:
```diff
@@ -26,21 +34,54 @@ interface ContactPointHeaderProps {
-  const isReferencedByRegularPolicies = policies.some((ref) => ref.route.type !== 'auto-generated');
+  const showManagePermissions = showManageContactPointPermissions(selectedAlertmanager!, contactPoint);
+
+  const regularPolicyReferences = policies.filter((ref) => ref.route.type !== 'auto-generated');
```
We should not filter out autogenerated policies for calculating the count. Take in account that only admin roles will get the autogenerated policies from the response. And in that case, I believe, admin users should count this policy in the count, as when filtering.
If we use this filtered list of polices, then there is a mismatch between what we see in the contact point list, and what we see in the policy tree after clicking to the header:

<img width="189" alt="Screenshot 2024-09-19 at 16 55 49" src="https://github.com/user-attachments/assets/52c3ccc3-813b-4b86-867a-fbf843d95827">

<img width="932" alt="Screenshot 2024-09-19 at 16 55 40" src="https://github.com/user-attachments/assets/e82a628c-a9fd-46c9-bfde-cc331ca0bb98">

yuri-tceretian replied:
I do not think that auto-generated policies should be counted in metadata because :
1. the actual usage is covered by UsedByRules
2. neither the user nor the admin have control over those policies. They are displayed for debugging purposes. 

However, I think we should not highlight them in the search results.

soniaAguilarPeiron replied:
> I do not think that auto-generated policies should be counted in metadata

I agree 💯

But If we don't highlight auto-generated policies in the filter, admin users are not able to see how many alert instances (the actual total)  are being routed to a particular contact point.
Maybe we could change the text in the contact point header to something like:  `Used by one 'regular' notification policy` ?

JacobsonMT replied:
If this is causing confusion for users, I would lean toward not showing the autogenerated policies in the UI at all, even to admins. IMO, they are an implementation detail that we decided might be useful for admins to see. 

The fact that we create root policies for contact points that aren't used by any rules is double-so an implementation detail that exists only to prevent accidental routing notifications to the default route.

konrad147 replied:
Maybe we should put it behind a feature toggle, if we/users needed it for debugging purposes? Or maybe instead a feature toggle we could add a query parameter that would reveal the autogenerated branch in the UI?

tomratcliffe replied:
Based on the discussion the other day in the experience weekly, how do we want to proceed here?

Right now, I've made it so, with the feature toggle off, it behaves as before. 

With it turned on, the number of policies shown only counts those that directly mention the contact point (i.e. excluding the auto generated ones for simplified routing).

It will also show the number of rules that explicitly mention the rule via simplified routing.

Are we okay to proceed with this and to tackle the autogenerated policy display separately? WDYT?

soniaAguilarPeiron replied:
maybe in case of `admin` user, we can render 'used by x notification polices (autogenerated are not included) or something like this?

konrad147 replied:
Let's keep it as it is and address our concerns in a follow-up PRs

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/ContactPointHeader.tsx:
```diff
@@ -26,21 +34,54 @@ interface ContactPointHeaderProps {
+  );
+
+  /** Number of rules that use this contact point for simplified routing */
+  const numberOfRules = Number(contactPoint.metadata?.annotations?.[ANNOTATION_INUSE_RULES]) || 0;
```
are we going to show zero rules in case we don't get this `contactPoint.metadata?.annotations?.[ANNOTATION_INUSE_RULES]`?

tomratcliffe replied:
If this ends up being 0 then the text won't show in the header at all, as we have a falsy check there :) unless I'm misunderstanding your question?

soniaAguilarPeiron replied:
sorry, you're right!, I missed this [line](https://github.com/grafana/grafana/pull/92885/files#diff-bd80acfd65ce0ebbe81500cff314e3c907861d173d3268a0ff86e532ea48f55fR191)

tomratcliffe committed:
Split contact point permissions checks for read/modify

JacobsonMT committed:
Cleanup permissions on config api reset and restore

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.test.tsx:
```diff
@@ -22,6 +25,16 @@ import setupVanillaAlertmanagerFlavoredServer, {
+ * Flushes out microtasks so we don't get warnings from @floating-ui/react
+ * as per https://floating-ui.com/docs/react#testing
+ */
+const flushMicrotasks = async () => {
```
maybe we could move this function to an utility folder? I guess we would need it in several places?

tomratcliffe replied:
I've moved this to a utils file and combined it with the other `testWithFeatureToggles` method etc that I added

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -173,11 +202,13 @@ const ContactPointsPageContents = () => {
-            active={showNotificationTemplates}
-            onChangeTab={() => setActiveTab(ActiveTab.NotificationTemplates)}
-          />
+          {showTemplatesTab && (
```
💯

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/contact-points/useContactPoints.tsx:
```diff
@@ -110,7 +110,8 @@ const useK8sContactPoints = (...[hookParams, queryOptions]: Parameters<typeof us
       return {
         ...result,
-        data,
+        // K8S API will return 403 error for no-permissions case, so its cleaner to fallback to empty array
```
💯

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/receivers/form/GrafanaReceiverForm.tsx:
```diff
@@ -160,6 +165,9 @@ export const GrafanaReceiverForm = ({ contactPoint, readOnly = false, editMode }
         commonSettingsComponent={GrafanaCommonChannelSettings}
         customValidators={{ [ReceiverTypes.OnCall]: onCallFormValidators }}
+        canManagePermissions={
+          editMode && contactPoint && showManageContactPointPermissions(GRAFANA_RULES_SOURCE_NAME, contactPoint)
```
why we don't allow editing permissions while creating a new contact point? is this intentional?

tomratcliffe replied:
We can't modify the permissions before creation as we don't have an ID to use for the access-control APIs

konrad147 commented on public/app/features/alerting/unified/utils/k8s/utils.ts:
```diff
@@ -30,3 +31,33 @@ type Entity = {
+export const canDeleteEntity = (k8sEntity: EntityToCheck) => getContactPointPermission(k8sEntity, 'canDelete');
+
+/** Annotation key that indicates how many notification policy routes are using this entity */
+export const ANNOTATION_INUSE_ROUTES = 'grafana.com/inUse/routes';
+
+/** Annotation key that indicates how many alert rules are using this entity */
+export const ANNOTATION_INUSE_RULES = 'grafana.com/inUse/rules';
```
What do you think about grouping known annotation keys in a one enum object?
Like
```ts
enum K8sAnnotations {
	InUseRoutes = 'grafana.com/inUse/routes',
	InUseRules = 'grafana.com/inUse/rules'
}

I feel it will be clear to group them together than keeping them as separate consts. Probably `provenance` could also fit there

tomratcliffe replied:
I like it :+1: Will make this change

tomratcliffe replied:
I made changes - but didn't entirely move `provenance` over as I didn't want to change the other uses and grow this PR any further. I marked the old annotation as deprecated so the few extra places will be more easily spotted and cleaned up after

konrad147 commented on public/app/features/alerting/unified/components/permissions/ManagePermissions.tsx:
```diff
@@ -0,0 +1,73 @@
+  );
+};
+
+export default ManagePermissions;
```
How about switching to a named export? Default exports can be harder for IDEs to import, tricky to navigate, and a bit confusing if renamed in the import. Named exports are usually clearer and easier to work with

tomratcliffe replied:
Sure, I can change it :+1: I think that we should work out some style/linting rules to apply (at least to the alerting code) so we're consistent about this kind of thing across the codebase.

konrad147 commented on public/app/features/alerting/unified/components/contact-points/constants.ts:
```diff
@@ -1,3 +1,28 @@
+ *
+ * Any permission in this list will be checked for client side access to view Contact Points functionality.
+ */
+const PERMISSIONS_CONTACT_POINTS_READ = [AccessControlAction.AlertingReceiversRead];
+
+/**
+ * List of granular permissions that allow modifying contact points
+ */
+export const PERMISSIONS_CONTACT_POINTS_MODIFY = [
+  AccessControlAction.AlertingReceiversCreate,
+  AccessControlAction.AlertingReceiversWrite,
+];
+
+/**
+ * List of all permissions that allow contact points read/write functionality
+ *
+ * Any permission in this list will also be checked for whether the built-in Grafana Alertmanager is shown
+ * (as the implication is that if they have one of these permissions, then they should be able to see Grafana AM in the AM selector)
+ */
+export const PERMISSIONS_CONTACT_POINTS = [...PERMISSIONS_CONTACT_POINTS_READ, ...PERMISSIONS_CONTACT_POINTS_MODIFY];
```
How about moving this to a file like `/contact-points/permissions`? We could also move the showManageContactPointPermissions function from the utils file there. I think keeping related functionality together in dedicated files would make things easier to maintain.

Based on my experience, files like `misc` and `utils` tend to become obsolete quickly and lead to code duplication over time. (no one really cares about these files)

tomratcliffe replied:
:+1: Have moved this to a `permissions` file

tomratcliffe committed:
Generalise getting annotation values from k8s entities

JacobsonMT committed:
Cleanup permissions on config api POST

note this is still not available with feature flag enabled

JacobsonMT committed:
Gate the permission manager behind FF until initial migration is added

konrad147 commented on public/app/features/alerting/unified/utils/k8s/constants.ts:
```diff
@@ -1,5 +1,24 @@
 /** Value of {@link PROVENANCE_ANNOTATION} given for entities that were not provisioned */
 export const PROVENANCE_NONE = 'none';
+
+export enum K8sAnnotations {
```
❤️

JacobsonMT committed:
Merge branch 'main' into alerting/contact-point-permissions

# Conflicts:
#	pkg/services/accesscontrol/ossaccesscontrol/receivers.go
#	pkg/services/ngalert/accesscontrol.go
#	pkg/services/ngalert/api/api_notifications_test.go
#	pkg/services/ngalert/ngalert.go
#	pkg/services/ngalert/notifier/alertmanager_config.go
#	pkg/services/ngalert/notifier/receiver_svc.go
#	pkg/services/ngalert/notifier/receiver_svc_test.go
#	pkg/services/ngalert/provisioning/contactpoints.go
#	pkg/services/ngalert/provisioning/contactpoints_test.go
#	pkg/services/provisioning/provisioning.go
#	public/locales/en-US/grafana.json
#	public/locales/pseudo-LOCALE/grafana.json

JacobsonMT committed:
Sync changes from config api PR

tomratcliffe committed:
Switch to named export

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/contact-point-permissions

JacobsonMT committed:
Merge branch 'main' into alerting/contact-point-permissions

JacobsonMT committed:
Revert unnecessary changes

JacobsonMT committed:
Revert Filter auth change and implement in k8s api only

JacobsonMT committed:
Don't allow new scoped permissions to give access without FF

Prevents complications around mixed support for the scoped permissions causing
oddities in the UI.

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPoints.tsx:
```diff
@@ -41,10 +46,16 @@ const ContactPointsTab = () => {
 
+  // If we're using the K8S API, then we don't need to fetch the policies info within the hook,
+  // as we get metadata about this from the API
+  const fetchPolicies = !shouldUseK8sApi(selectedAlertmanager!);
+  // User may have access to list contact points, but not permission to fetch the status endpoint
+  const fetchStatuses = contextSrv.hasPermission(AccessControlAction.AlertingNotificationsRead);
```
We aren't able to do that, as sometimes we don't want to fetch policies or statuses regardless - e.g. when showing the list of contact points in a dropdown, we never need that information

tomratcliffe committed:
Move to `permissions` file

JacobsonMT committed:
Fix integration tests to account for list permission change

tomratcliffe commented on public/app/features/alerting/unified/components/contact-points/ContactPointHeader.tsx:
```diff
@@ -26,21 +29,54 @@ interface ContactPointHeaderProps {
+  /** Is the contact point referenced by anything such as notification policies or as a simplified routing contact point? */
+  const isReferencedByAnything = numberOfRules + numberOfPolicies > 0;
+
+  /** Does the current user have permissions to edit the contact point? */
+  const hasAbilityToEdit = canEditEntity(contactPoint) || editAllowed;
+  /** Can the contact point be edited via the UI? */
+  const contactPointIsEditable = !provisioned;
+  /** Given the alertmanager, the user's permissions, and the state of the contact point - can it actually be edited? */
+  const canEdit = editSupported && hasAbilityToEdit && contactPointIsEditable;
 
-  const canEdit = editSupported && editAllowed && !provisioned;
-  const canDelete = deleteSupported && deleteAllowed && !provisioned && !isReferencedByRegularPolicies;
+  /** Does the current user have permissions to delete the contact point? */
+  const hasAbilityToDelete = canDeleteEntity(contactPoint) || deleteAllowed;
+  /** Can the contact point actually be deleted, regardless of permissions? i.e. ensuring it isn't provisioned and isn't referenced elsewhere */
+  const contactPointIsDeleteable = !provisioned && !isReferencedByAnything;
+  /** Given the alertmanager, the user's permissions, and the state of the contact point - can it actually be deleted? */
+  const canBeDeleted = deleteSupported && hasAbilityToDelete && contactPointIsDeleteable;
```
I've added some more tests now, and the coverage of the suite of tests with `alertingApiServer` enabled is at the following - is that okay? WDYT?

```
96.87% Statements
96.49% Branches
77.77% Functions 
98.36% Lines
```

tomratcliffe committed:
Add additional tests for contact points

tomratcliffe committed:
Fix redirect for viewer on edit page

tomratcliffe committed:
Combine alerting test utils and move to new file location

JacobsonMT committed:
Allow new permissions to access provisioning export paths with FF

tomratcliffe committed:
Always allow exporting if its grafana flavoured

yuri-tceretian reviewed (commented):
Backend code LGTM!

tomratcliffe committed:
Fix logic for showing auto generated policies

tomratcliffe committed:
Fix delete logic for contact point only referenced by a rule

konrad147 reviewed (approved):
LGTM!

tomratcliffe committed:
Suppress warning message when renaming a contact point

JacobsonMT committed:
Clear team and role perm cache on receiver rename

Prevents temporarily broken UI permissions after rename when a user's source of
elevated permissions comes from a cached team or basic role permission.

JacobsonMT committed:
Debug log failed cache clear on CopyPermissions
